File: patterns/architectural/layered-architecture.js
/**
 * Layered Architecture Pattern - JavaScript
 * 
 * Organizes code into horizontal layers, each with specific responsibilities:
 * - Presentation Layer: UI
 * - Business Layer: Business logic
 * - Data Access Layer: Database operations
 */

// Data Access Layer
class UserDAO {
    constructor() {
        this.users = [];
    }
    
    save(user) {
        this.users.push(user);
        return user;
    }
    
    findById(id) {
        return this.users.find(u => u.id === id);
    }
    
    findAll() {
        return this.users;
    }
}

// Business Layer
class UserService {
    constructor(userDAO) {
        this.userDAO = userDAO;
    }
    
    createUser(name, email) {
        // Business logic
        if (!this.isValidEmail(email)) {
            throw new Error('Invalid email');
        }
        
        const user = {
            id: Date.now(),
            name,
            email,
            createdAt: new Date()
        };
        
        return this.userDAO.save(user);
    }
    
    getUser(id) {
        return this.userDAO.findById(id);
    }
    
    getAllUsers() {
        return this.userDAO.findAll();
    }
    
    isValidEmail(email) {
        return email.includes('@');
    }
}

// Presentation Layer
class UserController {
    constructor(userService) {
        this.userService = userService;
    }
    
    createUser(req) {
        try {
            const user = this.userService.createUser(req.name, req.email);
            return { status: 200, data: user };
        } catch (error) {
            return { status: 400, error: error.message };
        }
    }
    
    getUser(id) {
        const user = this.userService.getUser(id);
        return user ? { status: 200, data: user } : { status: 404, error: 'User not found' };
    }
}

// Example usage
console.log('=== Layered Architecture Pattern ===\n');

const dao = new UserDAO();
const service = new UserService(dao);
const controller = new UserController(service);

const result = controller.createUser({ name: 'John Doe', email: 'john@example.com' });
console.log('Create user:', result);

const user = controller.getUser(result.data.id);
console.log('Get user:', user);

module.exports = { UserDAO, UserService, UserController };

File: patterns/architectural/layered-architecture.py
"""
Layered Architecture Pattern - Python

Organizes code into horizontal layers, each with specific responsibilities:
- Presentation Layer: UI
- Business Layer: Business logic
- Data Access Layer: Database operations
"""


# Data Access Layer
class UserDAO:
    def __init__(self):
        self.users = []
    
    def save(self, user):
        self.users.append(user)
        return user
    
    def find_by_id(self, id):
        return next((u for u in self.users if u['id'] == id), None)
    
    def find_all(self):
        return self.users


# Business Layer
class UserService:
    def __init__(self, user_dao):
        self.user_dao = user_dao
    
    def create_user(self, name, email):
        # Business logic
        if not self.is_valid_email(email):
            raise ValueError("Invalid email")
        
        import time
        user = {
            'id': int(time.time()),
            'name': name,
            'email': email,
            'created_at': time.time()
        }
        
        return self.user_dao.save(user)
    
    def get_user(self, id):
        return self.user_dao.find_by_id(id)
    
    def get_all_users(self):
        return self.user_dao.find_all()
    
    def is_valid_email(self, email):
        return '@' in email


# Presentation Layer
class UserController:
    def __init__(self, user_service):
        self.user_service = user_service
    
    def create_user(self, req):
        try:
            user = self.user_service.create_user(req['name'], req['email'])
            return {'status': 200, 'data': user}
        except ValueError as error:
            return {'status': 400, 'error': str(error)}
    
    def get_user(self, id):
        user = self.user_service.get_user(id)
        return {'status': 200, 'data': user} if user else {'status': 404, 'error': 'User not found'}


# Example usage
if __name__ == "__main__":
    print("=== Layered Architecture Pattern ===\n")
    
    dao = UserDAO()
    service = UserService(dao)
    controller = UserController(service)
    
    result = controller.create_user({'name': 'John Doe', 'email': 'john@example.com'})
    print("Create user:", result)
    
    user = controller.get_user(result['data']['id'])
    print("Get user:", user)

